### PR TITLE
Update section 11.7.3 to be more accurate and specific about expression tree conversions

### LINKED ISSUE #341: Strange reference to Section 12.7.3 in §C.3 Expression<TDelegate>.Compile()
In draft-v6, Section C.3 "Standard Library Types not defined in ISO/IEC 23271" includes this reference to Section 12.7.3: <https://github.com/dotnet/csharpstandard/blob/331aae414c0955eb460bd1cbdbcb94792ba83845/standard/standard-library.md#L568-L577>

However, [12.7.3 Simple names](https://github.com/dotnet/csharpstandard/blob/331aae414c0955eb460bd1cbdbcb94792ba83845/standard/expressions.md#1273-simple-names) does not mention "delegate". Perhaps the reference should be to some other section.

### PR DESCRIPTION
**Pull request by jskeet**
Expression tree conversions are implementation-defined, so it
doesn't make sense to refer to 11.7.3 for more details (or to fail
to do so properly due to using the wrong number, as is the case in
the standard library).

Fixes #341

### CODE REVIEW

**Comment by jskeet** [comment]
Meeting feedback: restore the references, but state in the text around the reference that it's implementation-specific, so that people don't *expect* more details.

**Comment by jskeet** [comment]
We discussed whether we could fix the ugliness of "the conversion exists but may fail", but decided not to.

**Comment by jskeet** [comment]
This is ready for another look in the meeting later.

**Comment by jskeet** [comment]
This is now *not* ready, as I've messed up with encodings. Jon to reapply the desired changes in a suitably UTF-8 aware editor.

**Comment by jskeet** [comment]
Looks like it was just a single character that was wrong - but that confused GitHub into thinking it wasn't UTF-8 after all.

Fixed in a third commit.

I can't remember whether we actually approved this or not, so I won't merge yet, but let's try to get it done next time.

**Review by Nigel-Ecma — CHANGES_REQUESTED** [review]
The motivation for this PR is good but unfortunately I don't think it's a good idea as the issue is more involved.

§11.7.1 starts:

> An *anonymous_method_expression* or *lambda_expression* is classified as an anonymous function (§12.16). The expression does not have a type, but can be implicitly converted to a compatible delegate type. Some lambda expressions may also be implicitly converted to a compatible expression-tree type.

Could be read as limiting expression-tree conversions to a (subset of) *lambda* expressions, what about anonymous method expressions?

Later in the same clause we get:

> A lambda expression `F` is compatible with an expression tree type` Expression<D>` if `F` is compatible with the delegate type `D`. This does not apply to anonymous methods, only lambda expressions.

> Certain lambda expressions cannot be converted to expression tree types: Even though the conversion *exists*, it fails at compile-time.  This is the case if the lambda expression:
> - Has a block body
> - Contains simple or compound assignment operators
> - Contains a dynamically bound expression
> - Is async

Here it is still talking only about lambda expressions and gives an explicit list of those that cannot be converted to expression tree types. Is the list exhaustive? Could possibly be read either way.

Which gets us to:

> ### 11.7.3 Evaluation of anonymous function conversions to expression tree types

> Conversion of an anonymous function to an expression-tree type produces an expression tree (§9.6). More precisely, evaluation of the anonymous-function conversion produces an object structure that represents the structure of the anonymous function itself. The precise structure of the expression tree, as well as the exact process for creating it, are implementation-defined.

while the PR replaces the last sentence with:

> The following aspects are implementation-defined:
> - The precise structure of the expression tree
> - The exact process for creating it
> - The rules for which lambda expressions can be successfully converted into expression trees, and for which expression tree types.

The original talks about anonymous functions, says the process/result is implementation defined and does not repeat the list above of those lambda expressions which are not convertible.

The PR version contradicts the previous list by saying that those that are not convertible is implementation defined. 

Overall in the original §11.7 there is a lack of clarity of what the rules are for anonymous functions (which the Standard is using to cover both *anonymous_method_expression* and *lambda_expression*) and what rules, if any, apply only to lambda expressions. The original provides a list of lambda expressions which cannot be converted, but it is unclear whether this is exhaustive; the PR contradicts the list by defining it as implementation defined.

Moving now to §9.6, para 2 starts:

> If a conversion exists from a lambda expression to a delegate type `D`, a conversion also exists to the expression tree type `Expression<TDelegate>`.

Seems we're back to lambda expressions, not §11.7.3’s anonymous functions. An x-ref to §11.7(*.x*?) might be helpful.

> Expression trees are efficient in-memory data representations of lambda expressions and make the structure of the lambda expression transparent and explicit.

Definitely *lambda expressions*, and yet §11.7.3 is about conversion of anonymous functions. (BTW “efficient” seems a bit odd as well, unless “implementation defined” is *required* to be efficient.)

And finally we come to:

> - Not all lambda expressions can be converted to expression trees. For instance, lambda expressions with statement bodies, and lambda expressions containing assignment expressions cannot be represented. In these cases, a conversion still exists, but it will fail at compile-time. These exceptions are detailed in §11.7.3).

(The unbalanced “)” is in the original.)

Here it does seem to suggest that the subset of lambda’s which cannot be convert is exhaustive (”are detailed”), but that list is in §11.7.1 not §11.7.3.

The PR changes this bullet to:

> - Not all lambda expressions can be converted to expression trees. For instance, lambda expressions with statement bodies, and lambda expressions containing assignment expressions cannot be represented. In these cases, a conversion still exists, but it will fail at compile-time. The conversion's existence is specified in §11.7.3), but this does not provide details due to the implementation-dependent nature of the conversion.

(The lone “)” survived).

Which moves us back to implementation-defined...

The stated aim of the PR is to remove an x-ref to a list which doesn’t exist, but a list does exist – just in §11.7.1 not §11.7.3.

There is some lack of clarity in the original between:

- what applies to anonymous function and what only applies to lambda expressions; and
- what is implementation-defined and what is not

The issues are larger than the PR tackles, so though the motivation is good I wouldn't recommend adopting it but rather aiming for a PR which tackles all the issues.

**Comment by jskeet** [comment]
(The quotes are from the above comment... hopefully it's intelligible that way.)

> Could be read as limiting expression-tree conversions to a (subset of) lambda expressions, what about anonymous method expressions?

This is deliberate. There is no conversion from anonymous method expressions to expression trees. It really is only lambda expressions, despite the heading of "anonymous function conversions". We should discuss (in a meeting) whether it's worth changing the headings here to be more explicit about only supporting lambda expressions.

> Here it is still talking only about lambda expressions and gives an explicit list of those that cannot be converted to expression tree types. Is the list exhaustive? Could possibly be read either way.

No, the list isn't exhaustive. We should fix it to be explicit that this is *not* exhaustive - or omit it entirely. It's not obvious to me whether *everything* should be implementation-defined, or whether it's useful for the standard to say "The following restrictions definitely apply, and implementations can impose additional restrictions".

I'll fix the closing parenthesis now - thanks for spotting that - but leave the rest of the PR right now; we can discuss the approach we want to take in the next meeting, and I'll update the PR afterwards.

**Comment by jskeet** [comment]
In our last meeting we wanted to know whether a conversion from "anonymous function" to expression tree existed, or only a conversion from "lambda expression". Experimentally, it looks like it's only from lambda expression:

```csharp
using System.Linq.Expressions;

int a = 0;
var derived = new DerivedClass();
derived.Method(delegate { return 5; }); // Fine
derived.Method(() => 5); // Fine
derived.Method(() => a = 5); // Fails
derived.Method(() => { return 5; }); // Fails

public class BaseClass
{
    public void Method(Func<int> func) { }
}

public class DerivedClass : BaseClass
{
    public void Method(Expression<Func<int>> expression) { }
}
```

The last two calls fail because the conversion exists but fails; the anonymous method call succeeds because the method in `DerivedClass` isn't applicable.

Next step is to make that clear in the text :)

**Comment by jskeet** [comment]
This is now updated - PTAL, particularly @Nigel-Ecma who had lots of comments before. I'm sure I've missed something.

File: standard/conversions.md
Comment (Nigel-Ecma) [inline standard/conversions.md]: This reads to be as out-of-order – the examples in lines 757-761 come _before_ the list of things in lines 767-769 of things which are implantation specific. Surely it should be the other way around?

File: standard/types.md
Comment (Nigel-Ecma) [inline standard/types.md]: Starting back at line 556 (I can't attach a comment there) it states that expression trees, an implementation defined type, "are efficient" – maybe "shall be", "must be", "are required to be", but surely not "are"? It carries on to state that they make "make the structure of the lambda expression transparent and explicit" – again this is an implementation defined type, so maybe "shall make…", "must make…" etc but not simply "and make".

In this section we also see repetition of material from §11.7 without at least x-refs, and maybe less repetition is possible.

I think this is proving a hard feature to Standardise because it is essentially implementation defined and the Standard is trying to be implementation neutral but hasn't much to work with, while also implementation specific material (e.g. it being efficient above) has crept in.

**Review by Nigel-Ecma — CHANGES_REQUESTED** [review]
I don't know *what* changes, sorry, but this still doesn't feel right – see comments.

File: standard/conversions.md
Comment (jskeet) [reply standard/conversions.md]: I went back and forth on this. I don't mind whether we go from the detail to the summary or vice versa. Let's talk about it in the meeting.

File: standard/types.md
Comment (jskeet) [reply standard/types.md]: Agree that we shouldn't have efficient in there - removed.

I've reworked this section a bit, restoring the cross-reference (but moving it) and removing duplication. I've retained the part about the `Compile` method, but it's no longer in a list.

Comment (jskeet) [reply standard/types.md]: Also removed the bit about return types and parameters which seemed to be unnecessary.

**Review by jskeet — COMMENTED** [review]
Updated - PTAL. I've left the changes in a second commit; I'll rebase them down to one when we're happy.

File: standard/conversions.md
Comment (Nigel-Ecma) [inline standard/conversions.md]: Not even the imprecise structure of the tree is defined, it may not even be a tree surely? `Expression<>` is an implementation defined type, the implementation provides a compile time conversion (so not a conversion operator? Or a constructor? Just a compilation strategy so not really a “conversion” *per se*?) which results in an opaque value of type `Expression<>`. And there is just one required method on this type `Compile`.

File: standard/types.md
Comment (Nigel-Ecma) [inline standard/types.md:567]: I think this particular deletion removes all reference to `Expression<>` being (largely) implementation-defined for the types.md, it is mentioned in (the later) conversions.md (see comment there).

File: standard/types.md
Comment (Nigel-Ecma) [inline standard/types.md]: I’m not sure “offers an” is right here, `Compile` is the only method the otherwise implementation-defined `Expression<>` type is *required* to provide.

File: standard/types.md
Comment (Nigel-Ecma) [inline standard/types.md]: `Expression<TDelegate>`'s internals are implementation defined, as far the Standard goes they are *opaque* and not “transparent and explicit”.

This doesn’t mean a particular implementation may not choose to surface the internals, that is allowed by the Standard.

**Review by Nigel-Ecma — CHANGES_REQUESTED** [review]
Apologies but I think we’re still not there on this one.

As far as I can see we are standardising *only*:

- There is a type `Expression<TDelegate>`
- Instances of it are constructed/converted/compiled from a lambda expression at compile time in an implementation defined way.
   - This construction/conversion/compilation may fail at compile time for an implementation defined reason  
- It provides a single method `Compile()` which returns a `TDelegate`
   - This method may fail at run time for an implementation defined reason  

And nothing else.

Unless of course the intention is to Standardise more?

File: standard/conversions.md
Comment (jskeet) [reply standard/conversions.md]: I've now removed this list anyway, due to other changes.

File: standard/conversions.md
Comment (jskeet) [reply standard/conversions.md]: This is now gone, but there's new wording in types.md.

File: standard/types.md
Comment (jskeet) [reply standard/types.md]: Changed to just "provides". (I think it's reasonably clear that this means "the implementation is required to provide".)

File: standard/types.md
Comment (jskeet) [reply standard/types.md:567]: Right - I've moved some aspects back to types.md.

File: standard/types.md
Comment (jskeet) [reply standard/types.md]: I've rewritten all of this as a note. I don't even know what "transparent" was meant to be saying before.

**Review by jskeet — COMMENTED** [review]
I've taken yet another pass at this.

Note that I *don't* want to be in the business of specifying anything extra - if anyone thinks we should be requiring more aspects to be specified, we should discuss this. I'm just trying to make sure it's clearer what the limits of the specification are.

**Review by Nigel-Ecma — COMMENTED** [review]
I think all vestiges of implementation specific stuff are gone and just the bare bones that are being standardised remain – the process was long but the result is short and sweet :-)

I don't know if the notes at the ends of §9.6 & §11.7.3 are what we wish to say or not, I'll defer to the collective wisdom on that.

So this is an approval on the normative text, leaving the collective wisdom at the meeting to approve the notes :-)

**Review by Nigel-Ecma — APPROVED** [review]

**Pull request merged**
